editor/src/messages/portfolio/document/node_graph/node_graph_message_handler.rs

```diff
@@ -1956,6 +1989,88 @@ impl<'a> MessageHandler<NodeGraphMessage, NodeGraphMessageContext<'a>> for NodeG
 					responses.add(FrontendMessage::UpdateBox { box_selection })
 				}
 			}
+			NodeGraphMessage::UpdateLassoSelection => {
+				if let Some((lasso_selection_curr, _)) = &mut self.lasso_selection_curr {
+					// WARNING WARNING WARNING: this commented-out code is copy pasted from UpdateBoxSelection above and has not been edited for lasso
+					// The mouse button was released but we missed the pointer up event
+					// if ((e.buttons & 1) === 0) {
+					// 	completeBoxSelection();
+					// 	boxSelection = undefined;
+					// } else if ((e.buttons & 2) !== 0) {
+					// 	editor.handle.selectNodes(new BigUint64Array(previousSelection));
+					// 	boxSelection = undefined;
+					// }
+
+					let Some(network_metadata) = network_interface.network_metadata(selection_network_path) else {
+						log::error!("Could not get network metadata in UpdateBoxSelection");
+						return;
+					};
+
+					{
+						let node_graph_point = network_metadata
+							.persistent_metadata
+							.navigation_metadata
+							.node_graph_to_viewport
+							.inverse()
+							.transform_point2(ipp.mouse.position);
+
+						lasso_selection_curr.push(node_graph_point);
+					}
+
+					let lasso_selection_viewport: Vec<DVec2> = lasso_selection_curr
+						.iter()
+						.map(|selection_point| network_metadata.persistent_metadata.navigation_metadata.node_graph_to_viewport.transform_point2(*selection_point))
+						.collect();
+
+					let shift = ipp.keyboard.get(Key::Shift as usize);
+					let alt = ipp.keyboard.get(Key::Alt as usize);
+					let Some(selected_nodes) = network_interface.selected_nodes_in_nested_network(selection_network_path) else {
+						log::error!("Could not get selected nodes in UpdateBoxSelection");
+						return;
+					};
+					let previous_selection = selected_nodes.selected_nodes_ref().iter().cloned().collect::<HashSet<_>>();
+					let mut nodes = if shift || alt {
+						selected_nodes.selected_nodes_ref().iter().cloned().collect::<HashSet<_>>()
+					} else {
+						HashSet::new()
+					};
+					let all_nodes = network_metadata.persistent_metadata.node_metadata.keys().cloned().collect::<Vec<_>>();
```

Comment:
It is quite annoying to have to call collect here, due to the very poorly designed `node_click_targets` which requires mutable access to the `NodeNetworkInterface`. However I guess there is little that can be done unless rustc adds partial borrows or someone refactors the network interface.